### DISCUSSION TITLE
FastAPI with Pydantic 2 - constr has no effect on query parameters

### DISCUSSION THREAD

**Discussion by sschaub** [fastapi/fastapi#10002] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydantic import constr

app = FastAPI()

@app.get("/demo")
def test(blah: constr(min_length=2) = Query()):
    pass
```

### Description

- Open browser and navigate to /demo?blah=a
- Expected: 422 (minimum length: 2)
- Actual: 200

### Operating System

Windows

### FastAPI Versions

0.110.1

### Pydantic Versions

2.6.4

### Python Version

3.10.5

**Comment by sschaub** [comment]
@tiangolo, This issue is still present in FastAPI 0.104.1 with Pydantic 2.5.1. However, it works properly with FastAPI 0.104.1 with Pydantic 1.10.13, so it is related to Pydantic 2.

**Comment by hwalhwal** [comment, marked as answer]
Here's the solution:

```python
import typing
from fastapi import FastAPI, Path
from pydantic import StringConstraints

app = FastAPI()

Length2String = typing.Annotated[str, StringConstraints(min_length=2)

@app.get("/demo/{blah}")
async def test(blah: Length2String = Path(annotation=Length2String)) -> str:
    return blah

```

**Reply by sschaub** [reply]
Thank you, @hwalhwal. Your example works.

I also tested a variation, which also works:

```
@app.get("/demo/{blah}")
async def test(blah: str= Path(annotation=typing.Annotated[str, StringConstraints(min_length=2)])) -> str:
    return blah
``` 

However, the code is much more verbose than what works with Pydantic 1:

```
@app.get("/demo/{blah}")
async def test(blah: constr(min_length=2) = Path()) -> str:
    return blah
```
